File: go/problem_6.go
package main

import "fmt"

// xorNode is a node in an XOR linked list. both holds the XOR of the
// previous and next node indices (using slice indices as stand-in for pointers).
type xorNode struct {
	val  int
	both int // prev_idx XOR next_idx; -1 used as null sentinel
}

// xorList is an XOR doubly-linked list backed by a slice of nodes.
// Using a slice lets us simulate pointer arithmetic in a GC language.
type xorList struct {
	nodes []xorNode
	head  int // index of first node (-1 if empty)
	tail  int // index of last node  (-1 if empty)
}

func newXorList() *xorList {
	return &xorList{head: -1, tail: -1}
}

// add appends an element to the end of the list.
func (l *xorList) add(val int) {
	idx := len(l.nodes)
	prev := l.tail
	// new node's both = prev XOR null(-1 XOR -1 = 0, but we use -1^-1 trick)
	// both = prev_idx XOR next_idx; next is -1 (none)
	both := prev ^ -1 // equivalent to prev XOR null
	_ = both          // stored below
	l.nodes = append(l.nodes, xorNode{val: val, both: prev ^ -1})
	if l.tail != -1 {
		// update old tail: its both was (prev_of_tail XOR -1), now XOR in new idx
		l.nodes[l.tail].both ^= (-1 ^ idx) // replace null with idx
	}
	l.tail = idx
	if l.head == -1 {
		l.head = idx
	}
}

// get returns the node value at the given 0-based index.
func (l *xorList) get(index int) int {
	prev := -1
	curr := l.head
	for i := 0; i < index; i++ {
		next := prev ^ l.nodes[curr].both // XOR out prev to get next
		prev = curr
		curr = next
	}
	return l.nodes[curr].val
}

func problem6() {
	fmt.Println("Problem 6:")
	lst := newXorList()
	lst.add(1)
	lst.add(2)
	lst.add(3)
	lst.add(4)
	if lst.get(0) != 1 { panic("assertion failed") }
	if lst.get(1) != 2 { panic("assertion failed") }
	if lst.get(2) != 3 { panic("assertion failed") }
	if lst.get(3) != 4 { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}

File: c/problem_6.c
#include "problem_6.h"
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

/* XOR linked list backed by a flat array. -1 is the null sentinel. */
#define XOR_NULL (-1)
#define MAX_NODES 64

typedef struct { int val; int both; } XorNode;

typedef struct {
    XorNode nodes[MAX_NODES];
    int     count;
    int     head;
    int     tail;
} XorList;

static void xorListInit(XorList *l) {
    l->count = 0; l->head = XOR_NULL; l->tail = XOR_NULL;
}

/* add appends an element to the end of the list. */
static void xorListAdd(XorList *l, int val) {
    int idx  = l->count++;
    int prev = l->tail;
    l->nodes[idx].val  = val;
    l->nodes[idx].both = prev ^ XOR_NULL;
    if (l->tail != XOR_NULL)
        l->nodes[l->tail].both ^= (XOR_NULL ^ idx);
    l->tail = idx;
    if (l->head == XOR_NULL) l->head = idx;
}

/* get returns the value at the given 0-based index. */
static int xorListGet(XorList *l, int index) {
    int prev = XOR_NULL, curr = l->head;
    for (int i = 0; i < index; i++) {
        int next = prev ^ l->nodes[curr].both;
        prev = curr; curr = next;
    }
    return l->nodes[curr].val;
}

void problem6(void) {
    printf("Problem 6:\n");
    XorList lst;
    xorListInit(&lst);
    xorListAdd(&lst, 1); xorListAdd(&lst, 2);
    xorListAdd(&lst, 3); xorListAdd(&lst, 4);
    assert(xorListGet(&lst, 0) == 1);
    assert(xorListGet(&lst, 1) == 2);
    assert(xorListGet(&lst, 2) == 3);
    assert(xorListGet(&lst, 3) == 4);
    printf("All Tests Passed\n");
}
